Source/DafnyCore/Compilers/Compiler-java.cs

```diff
@@ -2334,7 +2334,10 @@ private class GenericArrayElementLvalue : ILvalue {
 
     protected string GetClassPath(string targetFilename) {
       var targetDirectory = Path.GetFullPath(Path.GetDirectoryName(targetFilename));
-      return "." + Path.PathSeparator + targetDirectory + Path.PathSeparator + Path.Combine(targetDirectory, "DafnyRuntime.jar");
+      var classpath = Environment.GetEnvironmentVariable("CLASSPATH");
+      // Note that the items in the CLASSPATH must habs absolute paths because the compilation is performed in a subfolder of where the command-line is executed
+      var cp = classpath != null && classpath.Length != 0 ? (Path.PathSeparator + classpath) : "";
+      return "." + Path.PathSeparator + targetDirectory + Path.PathSeparator + Path.Combine(targetDirectory, "DafnyRuntime.jar") + cp;
```

Comment:
Have you considered using `string.Join(Path.PathSeparator, ".", targetDirectory, Path.Combine(targetDirectory, "DafnyRuntime.jar"), classpath)` instead?

Reply:
OK - though it is slightly more complicated because one component is optional.

Reply:
> OK - though it is slightly more complicated because one component is optional.

Woops, sorry. Maybe `string.Join(Path.PathSeparator, new [] { ".", targetDirectory, Path.Combine(targetDirectory, "DafnyRuntime.jar"), classpath }.Where(s => !string.IsNullOrEmpty(s)))`

Reply:
Well - that verges on unreadable or not-quickly-understandable to me. Can we stick with the if-then-else?

Reply:
What makes it hard to read? I feel extracting a variable is an easy way to make long expression easier to read:
```c#
var paths = new [] { ".", targetDirectory, Path.Combine(targetDirectory, "DafnyRuntime.jar"), classpath };
return string.Join(Path.PathSeparator, paths.Where(s => !string.IsNullOrEmpty(s)));
```

But you can also do an if-then if you prefer:
```c#
var paths = new List<string> { ".", targetDirectory, Path.Combine(targetDirectory, "DafnyRuntime.jar") };
if (!string.IsNullOrEmpty(classpath)) {
  paths.Add(classpath);
}
return string.Join(Path.PathSeparator, paths);
```